src/charm.py

```diff
@@ -1,69 +1,159 @@
+            if not mongod_enabled:
+                logger.error("failed to enable mongod.service")
+                self.unit.status = BlockedStatus("failed to enable mongod.service")
+                return
+
+        try:
+            self._open_port_tcp(self._port)
+        except Exception:
+            self.unit.status = BlockedStatus("failed to open TCP port for MongoDB")
+            return
+
+        if not self.mongo.is_ready():
+            self.unit.status = WaitingStatus("Waiting for MongoDB Service")
+            event.defer()
+            return
+
+        # initialize replica set
+        self.unit.status = MaintenanceStatus("initializing MongoDB replicaset")
+        logger.debug("initalizing replica set")
+        try:
+            logger.debug("initialzing replica set for these IPs %s", self._unit_ips)
+            self.mongo.initialize_replica_set(self._unit_ips)
+            self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
+        except Exception as e:
+            logger.error("Error initializing replica sets in _on_start: error={}".format(e))
+            self.unit.status = BlockedStatus("failed to initialize replicasets")
+            return
+
+        self.unit.status = ActiveStatus("MongoDB started")
+
+    def _open_port_tcp(self, port: int) -> None:
+        """Open the given port.
+        
+        Args:
+            port: The port to open.
+        """
+        try:
+            check_call(["open-port", f"{port}/TCP"])
+        except subprocess.CalledProcessError as e:
+            logger.exception(f"failed opening port {port}", exc_info=e)
+            raise Exception(f"failed to open port {port}")
+
+    def _add_mongodb_org_repository(self) -> None:
+        """Adds MongoDB repo to container."""
         repositories = apt.RepositoryMapping()
 
         # Get GPG key
         try:
             key = urlopen("https://www.mongodb.org/static/pgp/server-5.0.asc").read().decode()
         except Exception as e:
-            logger.exception(f"failed to get GPG key", exc_info=e)
+            logger.exception("failed to get GPG key, reason: %s", e)
             self.unit.status = BlockedStatus("Failed to get GPG key")
             return
 
         # Add the repository if it doesn't already exist
-        if "https://repo.mongodb.org/apt/ubuntu" not in repositories:
+        repo_name = "deb-https://repo.mongodb.org/apt/ubuntu-focal/mongodb-org/5.0"
+        if repo_name not in repositories:
             try:
-                line = "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/5.0 muliverse"
+                line = "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/5.0 multiverse"
                 repo = apt.DebianRepository.from_repo_line(line)
                 # Import the repository's key
                 repo.import_key(key)
                 repositories.add(repo)
             except Exception as e:
```

Comment:
You know what's coming now ;-)

Reply:
Yes! Exception handling! `from_repo_line` doesn't raise any exceptions on its own, but looking at the functions it calls it seems like it could throw an InvalidSourceError or a ValueError. I'll handle those